### PR TITLE
Apply rules to implicitly created directories (Issue #85)

### PR DESCRIPTION
**Pull request by moedan**
Bugfix to apply rules to implicitly created intermediate directories.

See issue #85 for details.

### CODE REVIEW

**Review by ctron — APPROVED** [review]

**Comment by ctron** [comment]
I very much prefer this approach. I like the opt-in character of it. And I think it a reasonable concept to understand. Good idea!

I am not so sure about the naming. Naming is hard I know :) … And am not trying to push back on the current naming, just checking if we could come up with something better.

Basically it's a set of base directories we enable for automatic generation of intermediate directories, starting from the provided base, right?

**Comment by moedan** [comment]
Yes, naming isn't so easy. I'm open for suggestions. ;)

> Basically it's a set of base directories we enable for automatic generation of intermediate directories, starting from the provided base, right?

Yes, you're right.

---

While browsing through the source code and documentation, I came across the `directories` flag from the collect entry:

```
  <entry>
    <name>/usr/lib</name>
    <collect>
      <from>src/main/resources/lib</from>
      <directories>true</directories> <!-- make explicit directories -->
    </collect>
  </entry>
```

If the flag is active, the intermediate folders are created explicitly (no problem with my approach). But if the flag is inactive, intermediate folders are still generated if the paths correspond to the base folders of my extension.
Do you think this is a problem or could cause confusion?

**Comment by ctron** [comment]
> If the flag is active, the intermediate folders are created explicitly (no problem with my approach). But if the flag is inactive, intermediate folders are still generated if the paths correspond to the base folders of my extension.
> Do you think this is a problem or could cause confusion?

Hm, … looks like I didn't work with that plugin for a while. So yes, that would be correct approach when "scanning"/"collecting".

I am wondering if this now boils down to a documentation issue: If you scan, we have a flag for this. If you put files explicitly, you need to create directories yourself.

I would somehow still feel your extension useful. But indeed maybe confusing. I am open for suggestions.

**Comment by ctron** [comment]
I am wondering, what the improvement with this PR over the current situation? I am not trying to push back, but coming up with a good description of it. Like an example use case where things are much better now.

**Comment by moedan** [comment]
Sorry for my late reply. Something came up after I pushed my commit.

---

I had a bad feeling about overwriting the field `directories` from collect entry configuraiton. So changed my code to not changing the behavior of collect entry itself.

Current behavior for following example:
- The intermediate directories (including the target) of name-field `/opt/mycompany/myapp/a/b` are added explicitly to rpm package
- The directories from inside of collect are added dependent on `directories` flag to rpm package

```
<generateIntermediateDirectories>
    <baseDirecotry>/opt/mycompany/myapp</baseDirecotry>
</generateIntermediateDirectories>

<entry>
    <name>/opt/mycompany/myapp/a/b</name>
    <collect>
        <from>${project.basedir}/src/main/data</from> <!-- contains only one file /x/y/foobar -->
        <directories>false</directories> <!-- make explicit directories -->
    </collect>
    <ruleset>my-default</ruleset>
</entry>
```

Result with `directories=false`:

```
/opt/mycompany/myapp
/opt/mycompany/myapp/a
/opt/mycompany/myapp/a/b
/opt/mycompany/myapp/a/b/x/y/foobar
```

Result with `directories=true`:

```
/opt/mycompany/myapp
/opt/mycompany/myapp/a
/opt/mycompany/myapp/a/b
/opt/mycompany/myapp/a/b/x
/opt/mycompany/myapp/a/b/x/y
/opt/mycompany/myapp/a/b/x/y/foobar
```

I don't know  a real use case for this situation. Perhaps my extension should override the behavior of the dictionary property in collection entry after all (behavior before last commit).

**Comment by ctron** [comment]
Thanks for the explanation. So I think it's better to keep this independent of the `directories` flag, as it seems those are actually two independent settings.

The `directories` setting actually defaults to `true`, so it should automatically create directories.

The change you added would allow one to control the automatic generation of directories outside of the collector, right? I think that's a good feature. We just need to document that properly :) Otherwise it might get a bit confusing. And I think it might help having your example as part of this. I also think the naming is fine when it comes to that.

I know you've done a lot already, and this drag on a bit. But I would hope that you could do a final thing, add some docs around here: https://github.com/ctron/rpm-builder/blob/74ecb1a08d61d95617827faadfd844b51bbc6818/src/site/markdown/entry.md?plain=1#L52-L54

Maybe create a small new sections about auto-creation of directories at the end of the file, but also mention the difference between both concepts in this location linked. It doesn't need to be much, just to explain the difference of both mechanisms. That would be awesome.

**Review by ctron — DISMISSED** [review]
It's good in any case. Maybe we can add some quick docs.

**Comment by moedan** [comment]
Thanks for your positive feedback! I'll try to contribute the requested improvements to the documentation on Monday. Perhaps I will come up with a better name for the configuration parameter by then.

> The change you added would allow one to control the automatic generation of directories outside of the collector, right?

Yes, that's correct.

File: src/site/markdown/entry.md [hunk 1]
```diff
@@ -144,3 +147,72 @@ In order to use explicit information use the following elements:
     </entry>
 
 Also see [payload entry information](payload_information.html) for a full list.
+
+## Generate intermediate directories
+
+Since version `1.11.1 (TODO: enter correct version)` there is an option to automatically generate and add intermediate directories
```
Comment (ctron) [inline src/site/markdown/entry.md:153]: That should be the next minor version (major.minor.patch).

**Review by ctron — CHANGES_REQUESTED** [review]
This is awesome! Thanks.

The next version would be `1.11.0`, if you could replace that, I'll merge and try to get it out today.

**Pull request merged**